src/Session.cs

```diff
@@ -595,12 +597,38 @@ void OnDispose(Dispose dispose)
                             linkedList.Remove(delivery);
                         }
 
-                        delivery.OnStateChange(dispose.State);
+                        disposedDeliveries.Add(delivery);
                     }
 
                     delivery = next;
                 }
             }
+
+            // Update the state of the disposed deliveries
+            // Note that calling delivery.OnStateChange may complete some pending SendTask, thereby triggering the
+            // execution of some continuations. To avoid any deadlock, this MUST be done outside of any
+            // locks (cf issue https://github.com/Azure/amqpnetlite/issues/287).
+            // Also, to avoid delaying some tasks in case multiple deliveries are to be notified, we marshall all these
+            // notifications to new tasks, except the last one
+            for (int i = 0; i < disposedDeliveries.Count; i++)
+            {
+                var delivery = (Delivery)disposedDeliveries[i];
+                disposedDeliveries[i] = null;   // Avoid trailing reference
+#if !NETFX35 && !NETMF
+                if (i < disposedDeliveries.Count - 1)
+                {
+                    // Marshall the OnStateChange call to another task so that this one doesn't get held because
+                    // of potential continuations
+                    System.Threading.Tasks.Task.Factory.StartNew(() => delivery.OnStateChange(dispose.State));
```

Comment:
I prefer keeping the threading decision out of the session/link state machine implementation in the library. Here the session should simply notify delivery state change by directly calling delivery.OnStateChange. If the app needs to process the notification concurrently, it should fork in the async callback or task continuation. This is the pattern used for handling all protocol frames.

Reply:
I see your point, but I'm afraid calling directly `delivery.OnStateChange` might cause performance issues (or worse) when multiple threads/tasks are calling `SendAsync`.
Imagine a situation where unrelated tasks T1 and T2 call `SendAsync` and await on the result, and the corresponding delivery states get queued up. When T1 is awaken it has a long compute-bound operation to do; it cannot possibly know it should fork an async continuation to let T2 receive its result.

Reply:
That's a valid case. However, there is the other case where notification processing is minimal and the application wants the notification as fast as possible with minimal or no overhead. The application should know its usage pattern and make the decision accordingly. Another example is receive. When a message arrives should we deliver it to the callback directly or schedule it on thread pool? This can be extended to all places where user's callback is invoked.

I think its better to leave the decision to the application. If one pattern is common enough someone can build a layer on top of this library to expose APIs that target those applications specifically.

Reply:
> there is the other case where notification processing is minimal and the application wants the notification as fast as possible with minimal or no overhead

That should be mostly handled by the "else" case in this logic: we only run `TaskFactory.StartNew` if there are multiple pending deliveries to be notified, which should be pretty rare, shouldn't it? In most cases notifications will be delivered synchronously.

> The application should know its usage pattern and make the decision accordingly. 

That doesn't seem to me like a reasonable ask from client code. Surely as a library user I should be allowed to write such a piece of code:

```csharp
await senderLink.SendAsync(message);
CalculateLargeAmountOfData();
await otherAsyncOperation();
```

without having to worry whether that is going to stop other sender links from having their `.SendAsync()` call from completing?

And even worse, without the `TaskFactory.StartNew()` case the code above will "mostly" always work fine; until under heavy load delivery notifications will start queueing up, and then the sender task will start stopping each others from completing their tasks.

> Another example is receive. When a message arrives should we deliver it to the callback directly or schedule it on thread pool?

That's a very valid point. I haven't  checked whether similar issues can happen on the receiver side. But if so, then I think it should be addressed in a similar way. You shouldn't have one receiver hold up other receivers because its task continuation is being run.

Reply:
I see two issues here,
1. Deadlock due to invoking callback/task continuation from the session's lock. This can be fixed by doing so outside of the lock.
2. Invoke through direct call or scheduling. I understand your scenario but I am not convinced that we should do it at the session level. I don't consider it a common scenario where a session is shared by multiple links from multiple threads. If a user needs parallel processing and also wants isolation, it is better to use separate connections. From the amqp protocol perspective, a connection is single threaded (i.e. the protocol frames are processed in the order when they are received).
The most efficient way to send messages is to pipeline transfers. Most brokers batch the acks in one disposition in this case. The Task.StartNew change will cause context switch for most deliveries, even when the app is not using task at all. Also the fix does not completely solve the problem. The last delivery could be still blocking the connection's frame pump and delaying other frames.
I believe the price of context switching should be paid when it is needed, which is only known by the application. In your example, the developer knows that the middle function is heavy, so the code should be written like this.
```
await senderLink.SendAsync(message);
await Task.Yield();
CalculateLargeAmountOfData();
await otherAsyncOperation();
```
We could introduce extension methods that allow user to explicitly control continuation execution. For example,
```
        public Task SendAsync(this SenderLink sender, Message message, TimeSpan timeout, TaskContinuationOptions options)
        {
            return sender.SendAsync(message, timeout).ContinueWith(t => t, options).Unwrap();
        }
```

My suggestion is to address the deadlock issue in this pr. We can open a new issue for the task continuation issue and continue the discussion if needed.

Reply:
Ok that makes sense. Thanks for the detailed response.
I definitely agree we should tackle point 2. later. And you make a very good point that scheduling all but the last doesn't really fix the problem anyway, so I'll remove that from this PR.

Let's open a different issue as you suggest; I find it quite unfortunate to expect user code to rely on `Task.Yield` to ensure the message pump. That doesn't exactly fit the ["Pit of success"](https://blogs.msdn.microsoft.com/brada/2003/10/02/the-pit-of-success/) we would want for lib users. Maybe `SendAsync` should yield by default, with an option not to if the user knows what they're doing? Anyway, let's discuss this in an another issue.